src/algebra/jordan/basic.lean

```diff
@@ -0,0 +1,230 @@
+/-
+Copyright (c) 2021 Christopher Hoskin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+import algebra.ring.basic
+import algebra.lie.of_associative
+import data.real.basic
+import linear_algebra.basic
+
+/-!
+# Jordan algebras
+
+We define a set of conditions, formulated in terms of commuting multiplication operators, for the
+multiplication on a non-unital, non-associative semiring to be a Jordan multiplication. When the
+multiplication is commutative, these take a particularly simple form.
+
+A real Jordan algebra `A` can be introduced by
+```
+variables {A : Type*} [non_unital_non_assoc_ring A] [module ℝ A] [smul_comm_class ℝ A A]
+  [is_scalar_tower ℝ A A] [comm_jordan A]
+```
+
+## Main results
+
+- lin_jordan : Liniarisation of the commutative Jordan axiom
+
+## Implementation notes
+
+We shall primarily be interested in linear Jordan algebras (i.e. over rings of characteristic not
+two) leaving quadratic algebras to those better versed in that theory.
+
+The conventional way to linearise the Jordan axiom is to equate coefficients (more formally, assume
+that the axiom holds in all field extensions). For simplicity we use brute force algebraic expansion
+and substitution instead.
+
+## References
+
+* [Hanche-Olsen and Størmer, Jordan Operator Algebras][hancheolsenstormer1984]
+* [Cabrera García and Rodríguez Palacios, Non-associative normed algebras. Volume 1]
+  [cabreragarciarodriguezpalacios2014]
+-/
+
+set_option old_structure_cmd true
+
+/-- A not-necessarily-unital, not-necessarily-associative ring. -/
+@[protect_proj, ancestor add_comm_group non_unital_non_assoc_semiring ]
+class non_unital_non_assoc_ring (α : Type*) extends
+  add_comm_group α, non_unital_non_assoc_semiring α
+
+variables {A : Type*}
+
+/- A (unital, associative) ring is a not-necessarily-unital, not-necessarily-associative ring -/
+@[priority 100] -- see Note [lower instance priority]
+instance ring.to_non_unital_non_assoc_semiring (B : Type*) [_i : ring B] :
+  non_unital_non_assoc_ring B :=
+{ zero_mul := zero_mul,
+  mul_zero := mul_zero,
+  .. _i }
+
+variables [non_unital_non_assoc_ring A]
+
+namespace non_unital_non_assoc_ring
+-- For some reason `def L : A→+add_monoid.End A := add_monoid_hom.mul` doesn't work here?
```

Comment:
It's because `add_monoid_hom.mul` asks for a stronger typeclass than it needs. You should go weaken the typeclass there.

Reply:
Done: https://github.com/leanprover-community/mathlib/pull/11165